### DISCUSSION TITLE
support copy_file_range() syscall

### DISCUSSION THREAD

**Discussion by derekdai** [openzfs/zfs#4237] [Ideas]
This is a interesting new syscall first appeared in linux kernel 4.4 that offloaded file copy from user space to kernel, and I think this is a great way to implement a lightweight copying in between zfs file systems. Is there a plan to support it on ZoL?

**Comment by DeHackEd** [comment]
At the end of the day ZFS doesn't have anything to offer which "accelerates" the process. This system call merely removes the copies from user and kernel spaces. In most situations that's not going to be substantially faster than your disks or controllers can perform the IO.

**Reply by NobodyXu** [reply]
Wine now uses this syscall to deduplicate system files according to [here](https://www.phoronix.com/scan.php?page=news_item&px=Reflink-For-Wine-Patches), so it might be a good idea to support copy_file_range with reflink support.

**Comment by derekdai** [comment]
The storage won't go faster because of this syscall, but it saves resources from
- Prevent data from been copy to and from kernel and user space
- zfs can choose the best block size to copy to and from
- zfs don't need to recalculate checksum
- zfs don't need to decompress blocks
- zfs can do deduplication more efficiently

**Comment by lundman** [comment]
Also, macOS needs `vnop_clonefile` and `vnop_exchange` (which would like to use reflink) It is more about these operations being `atomic`. Either completes as a whole, or is cleaned up, with optional clone/exchange of xattrs, so keep that in mind if someone were to implement it :)

**Comment by baryluk** [comment]
FYI. copy_file_range is implemented on few Linux file systems, with varying level of optimisations:

- btrfs - uses CoW reflinking
- ext4 - reflinking extents, fallbacks to fast page cache copying
- xfs - ditto
- cifs - server side copy
- nfs - server side copy (and if server supports CoW on backend storage)
- fuse - whatever userspace app wants it to be
- overlayfs - delegates to upperdir

There were btrfs and xfs specific ioctls in the past to do similar, but now it is more standarized.

Would be nice to see this in ZFS somehow. As others mentioned, wine / proton uses this to reduce file duplication between wine prefixes.

This syscall is readily available in many programming languages, including C / glibc, Python 3.8, golang, Rust.

Also older versions of GNU coreutils `cp` used this btrfs ioctl, newer versions do utilize `copy_file_range` and do have it enabled by default for some scenarios (and fallback to other methods if needed i.e. running on very old kernel).

**Comment by rincebrain** [comment]
I have this wired up minimally successfully to the BRT PR, though there are quirks in that case when using it cross-filesystem due to Linux hardcoding restrictions on things.

Though I would point out that as FreeBSD recently encountered with their `copy_file_range` implementation, just because one would expect it to be faster to use doesn't make it so...

**Comment by ryao** [comment]
@rincebrain What are the quirks?

**Reply by rincebrain** [reply]
[See here](https://lists.gnu.org/archive/html/bug-coreutils/2022-05/msg00031.html) for where I asked coreutils about this behavior and they agreed it was wacky, but not something they could do anything about, and I don't think we can either.

The end of the BRT thread's current discussion points to Linux changing this behavior again, but my understanding from that discussion in the thread is that it should behave the same now for our purposes as it did before that change; I need to sync my tree and test that though...

...also clean up that tree and share it, but I wanted to script some robust stress testing before I did that.

**Reply by ryao** [reply]
We could bypass that by implementing a different `ioctl()`, but it would only be worth doing that if the coreutils developers would agree to add a fallback path that does a `statfs()` call to check for ZFS and then if it sees ZFS, attempts our custom ioctl.

**Reply by rincebrain** [reply]
I mean, I also have an implementation that uses that, but I can't really see coreutils agreeing to add a custom codepath for that.

**Reply by ryao** [reply]
SQLite has custom code paths for F2FS. I do not think coreutils adding code for ZFS would be that far fetched. It would only activate when EXDEV is returned, which is rare, and it would be a win for users. It would make sense to me. We are not close enough to having it ready that I would engage with them on it just yet though.

**Reply by rincebrain** [reply]
I don't think it's beyond the pale, but I would be surprised if the developers would agree to it given that nobody suggested it in the above discussion of quirks in behavior - and even if they did, I would doubt a lot of older distros would see the code backported, so we'd wind up needing something like `zfs cp` anyway...

**Reply by ryao** [reply]
Things like this take time to reach end users. It would be okay if not every distribution has the latest coreutils right away, provided that the hypothetical ZFS fallback path becomes a real one.

**Reply by rincebrain** [reply]
I'm not sure why you're describing it as hypothetical; PJD's PR has it implemented for FreeBSD, and my tree has it implemented for Linux.

**Reply by ryao** [reply]
I am referring to the coreutils patch.

**Reply by rincebrain** [reply]
Ah, I see.

I think a good historical question would be whether they ever supported the BTRFS-specific dedup ioctls,  or only supported it once FICLONE became more generic...

**Reply by jittygitty** [reply]
@rincebrain I think I definitely remember seeing the btrfs clone ioctl in an strace on cp --reflink before. Is that what you were asking? (In #13349 btrfs_ioc_clone: https://github.com/git-lfs/git-lfs/issues/3792 )

Is having separate "zfs cp" etc too problematic, until coreutils devs hook us up?

**Reply by ryao** [reply]
The btrfs IOCTLs were promoted to FICLONE and FICLONERANGE:

https://www.man7.org/linux/man-pages/man2/ioctl_ficlonerange.2.html

**Reply by ryao** [reply]
I would prefer to coordinate with the coreutils developers to get support for a ZFS specific ioctl into coreutils and then let support gradually make its way into distributions. It is better than the real possibility of doing nothing from being blocked on mainline.

If we were to make `zfs cp` people would expect us to keep it around forever. It is okay if cross mount reflinks do not work until the newer coreutils gets to users. It is likely a rare use case.

**Reply by rincebrain** [reply]
I personally would call that unacceptable, because one of my big use cases is that.

I can't find any evidence in quickly searching coreutils' git tree that they ever supported BTRFS_IOC_FILE_EXTENT_SAME and friends. Based on the previous exchange and that, I would suspect they would tell us to go work with Linux on a solution instead.

e: You know, I post that, and hit enter to close my scrolling, and one page later...

coreutils/coreutils@45330176690b079ed47ac7c58f29a1b028f97b07

**Reply by ryao** [reply]
> I can't find any evidence in quickly searching coreutils' git tree that they ever supported BTRFS_IOC_FILE_EXTENT_SAME and friends. Based on the previous exchange and that, I would suspect they would tell us to go work with Linux on a solution instead.

I think it would be better to ask them and get their opinion rather than deciding it ourselves. Otherwise, they could be willing to do it and we would not take advantage of that solely because we did not ask.

> I personally would call that unacceptable, because one of my big use cases is that.

You can install newer coreutils versions yourself in /usr/local. It is also not hard to make an out of tree tool for using it that we don't have to commit to maintaining ad infinitum.

**Reply by baryluk** [reply]
While I would also would love this supported, there is no point rushing it. It is probably better to clarify things with VFS maintainers, and even if it take a year or two to get a proper solution, than some `ioctl` hack, that is still a better solution.
